### ISSUE TITLE
Multi-level Navigation submenus appear outside viewport bounds when expanded

### ISSUE THREAD

**Issue by kishanjasani** [WordPress/gutenberg#48987]
### Description

**Navigation Block**: Block breaks the screen when someone adds multilevel navigation to the end of the screen.

See the below video for more information.

### Step-by-step reproduction instructions

1. Go to the header template.
2. Add Navigation Block
3. Add some nested navigation items( 2-3 level nesting ) on the right side( At the end of the screen ).
4. Horizontal scrollbar will be visible.

### Screenshots, screen recording, code snippet

Video: https://www.loom.com/share/1ae756ec792e4f8d97d24431c25181c4

### Environment info

**WordPress Vesion:**  6.1.1
**Gutenberg Version:** Trunk branch
**Active theme:**           Twenty Twenty Three

Browser: Chrome
OS: Macbook ( Ventura )

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by kathrynwp** [comment]
Thanks for the report, @kishanjasani. I think this is a common issue in LTR languages when sub-menu items open out to the right, rather than to the left. It was the same in classic themes. I'm flagging this for design feedback.

**Comment by annezazu** [comment]
Flagging for @getdave 's awareness as part of broader navigation efforts.

**Comment by getdave** [comment]
This is noted. We'll need JavaScript to solve this issue. It's probably going to become more of a priority once we start to explore mega menus.

**Comment by getdave** [comment]
@WordPress/gutenberg-components Are there any utilities we could make use of here for the Nav block such as we use for Popover that could be generically applied to keep the UI "in view"?

**Comment by mirka** [comment]
> [@WordPress/gutenberg-components](https://github.com/orgs/WordPress/teams/gutenberg-components) Are there any utilities we could make use of here for the Nav block such as we use for Popover that could be generically applied to keep the UI "in view"?

Not exactly sure if you need the same utility for frontend and backend, or what level of abstraction you're looking for, but some options are:

- [`Popover` (from `@wordpress/components`)](https://wordpress.github.io/gutenberg/?path=/docs/components-popover--docs) with the `shift` prop enabled (see also the `flip` prop, which is enabled by default) — I would look into this first if you want code deduplication on the backend
- [`Popover` from Ariakit](https://ariakit.org/components/popover) with the `shift`/`flip` props — Maybe better if it's frontend only
- [`Menu` from Ariakit](https://ariakit.org/components/menu) with the `shift`/`flip` props — Use this if you also want all the menu semantics and submenu logic
- [`floating-ui`](https://floating-ui.com/) with the `shift`/`flip` props — Lowest level utility, what all the above are built with

**Comment by ciampo** [comment]
I agree with what @mirka said — I would advise against using `@wordpress/components` for the front-end. Using a low-level library like `floating-ui` seems like the right solution.

**Comment by getdave** [comment]
Thank you both - thats excellent context 🙇

**Comment by westonruter** [comment]
I just noticed a worse outcome if I put the nav menu on the left in the header with the logo on the right. The nav menus expand left to a point where they cannot be seen at all:

> https://github.com/user-attachments/assets/129214b4-31ba-4d16-966f-e0e578cf54a9

This is how my submenus are configured:

> ![Image](https://github.com/user-attachments/assets/124d6189-3d41-4f7e-8024-340abf78eaf4)

I think the right solution for this is Anchor Positioning, as this avoids the need for any JavaScript. In particular, the `@position-try` directive in CSS allows you to position the submenu to the right (or left) if there is space, but otherwise to position the submenu below. See [example](https://developer.chrome.com/blog/anchor-positioning-api#adjust_anchor_positions_with_position-try).

Anchor Positioning is not yet available in all browsers, but it is at 71% of global users ([Can I Use](https://caniuse.com/css-anchor-positioning)), with it being implemented currently in Chromium and Safari TP. It's also part of Interop 2025, so Firefox should land support for it this year as well. In all, this could be a good progressive enhancement over the [current positioning CSS](https://github.com/WordPress/gutenberg/blob/1a7b8350ff4dd1ad028db2969a1588900c13f6dd/packages/block-library/src/navigation/style.scss#L32-L42).
